integration/spark2/src/main/scala/org/apache/spark/sql/events/MergeBloomIndexEventListener.scala

```diff
@@ -24,59 +24,88 @@ import org.apache.spark.internal.Logging
         val carbonTable = DataMapStoreManager.getInstance().getCarbonTable(carbonTableIdentifier)
-        val tableDataMaps = DataMapStoreManager.getInstance().getAllDataMap(carbonTable)
-        val sparkSession = SparkSession.getActiveSession.get
 
-        // filter out bloom datamap
-        var bloomDatamaps = tableDataMaps.asScala.filter(
-          _.getDataMapSchema.getProviderName.equalsIgnoreCase(
+        // filter out current rebuilt bloom datamap
+        val bloomDatamaps = DataMapStoreManager.getInstance().getAllDataMap(carbonTable).asScala
+          .filter(_.getDataMapSchema.getProviderName.equalsIgnoreCase(
             DataMapClassProvider.BLOOMFILTER.getShortName))
-
-        if (datamapPostEvent.isFromRebuild) {
-          if (null != datamapPostEvent.dmName) {
-            // for rebuild process
-            bloomDatamaps = bloomDatamaps.filter(
-              _.getDataMapSchema.getDataMapName.equalsIgnoreCase(datamapPostEvent.dmName))
-          }
-        } else {
-          // for load process, skip lazy datamap
-          bloomDatamaps = bloomDatamaps.filter(!_.getDataMapSchema.isLazy)
-        }
+          .filter(_.getDataMapSchema.getDataMapName.equalsIgnoreCase(datamapPostEvent.dmName))
+          .toList
 
         val segmentIds = datamapPostEvent.segmentIdList
-        if (bloomDatamaps.size > 0 && segmentIds.size > 0) {
-          // we extract bloom datamap name and index columns here
-          // because TableDataMap is not serializable
-          val bloomDMnames = ListBuffer.empty[String]
-          val bloomIndexColumns = ListBuffer.empty[Seq[String]]
-          bloomDatamaps.foreach( dm => {
-            bloomDMnames += dm.getDataMapSchema.getDataMapName
-            bloomIndexColumns += dm.getDataMapSchema.getIndexColumns.map(_.trim.toLowerCase)
-          })
-          new CarbonMergeBloomIndexFilesRDD(sparkSession, carbonTable,
-            segmentIds, bloomDMnames, bloomIndexColumns).collect()
-        }
+        mergeBloomIndex(sparkSession, carbonTable, bloomDatamaps, segmentIds)
     }
   }
 
 
-  private def clearBloomCache(carbonTable: CarbonTable, segmentIds: Seq[String]): Unit = {
-    DataMapStoreManager.getInstance.clearDataMaps(carbonTable.getTableUniqueName)
+  private def mergeBloomIndex(sparkSession: SparkSession, carbonTable: CarbonTable,
+      bloomDatamaps: List[TableDataMap], segmentIds: Seq[String]) = {
+    if (bloomDatamaps.nonEmpty && segmentIds.nonEmpty) {
+      // we extract bloom datamap name and index columns here
+      // because TableDataMap is not serializable
+      val bloomDMnames = ListBuffer.empty[String]
+      val bloomIndexColumns = ListBuffer.empty[Seq[String]]
+      bloomDatamaps.foreach(dm => {
+        bloomDMnames += dm.getDataMapSchema.getDataMapName
+        bloomIndexColumns += dm.getDataMapSchema.getIndexColumns.map(_.trim.toLowerCase)
+      })
+      LOGGER.info("Start to merge bloom index file")
```

Comment:
it's better to add the key message on the three logs, such as tablename , segmentids,bloomDatamaps

Reply:
fixed